### PR TITLE
support for Resnet50 V1.5

### PR DESCRIPTION
**Pull request by LakshmiKumar23**
- Modified argmax layer to support INT64 type ouput

- Modified anntest.cpp to support INT64 type output

- Modified tensor_matrix_multiply layer to support all combinations of 2D tensors

- Modified model_compiler for argmax and matmul

### CODE REVIEW

**Comment by kiritigowda** [comment]
@LakshmiKumar23 let me know when this is ready to be merged

**Comment by LakshmiKumar23** [comment]
@kiritigowda This is ready to be merged as soon as @rrawther approves it.

File: amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp [hunk 1]
```diff
@@ -83,10 +83,10 @@ static vx_status VX_CALLBACK validate(vx_node node, const vx_reference parameter
     ERROR_CHECK_STATUS(vxSetMetaFormatAttribute(metas[4], VX_TENSOR_DIMS, output_dims, sizeof(output_dims)));
 
     // check that tensors are 2D
-    if (input1_dims[3] != 1 || input1_dims[2] != 1 ||
-        input2_dims[3] != 1 || input2_dims[2] != 1 ||
-        input3_dims[3] != 1 || input3_dims[2] != 1 ||
-        output_dims[3] != 1 || output_dims[2] != 1)
+    if ((input1_dims[3] != 1 || input1_dims[2] != 1) && (input1_dims[1] != 1 || input1_dims[0] != 1) ||
+        (input2_dims[3] != 1 || input2_dims[2] != 1) && (input2_dims[1] != 1 || input2_dims[0] != 1) ||
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp]: this condition is ambiguous.
Please change is something like ((input1_dims[3]&input1_dims[2]) != 1) || ((input1_dims[1]&input1_dims[0]) != 1)

File: amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp [hunk 2]
```diff
@@ -95,16 +95,32 @@ static vx_status VX_CALLBACK validate(vx_node node, const vx_reference parameter
...
+            std::swap(input3_dims[0], input3_dims[1]);
+        }
+        else if (input3_dims[0] == 1 && input3_dims[1] == 1) {
+            std::swap(input3_dims[2], input3_dims[3]);
+        }
     }
-    if(input1_dims[0] != input2_dims[1] ||
-       input1_dims[1] != output_dims[1] || input2_dims[0] != output_dims[0] ||
+    if(((input1_dims[0] != input2_dims[1] ||
+       input1_dims[1] != output_dims[1] || input2_dims[0] != output_dims[0]) && (input1_dims[2] != input2_dims[1] ||
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp]: too many conditions, replace with better code

File: amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp [hunk 3]
```diff
@@ -161,9 +177,20 @@ static vx_status VX_CALLBACK initialize(vx_node node, const vx_reference *parame
     data->tA = params.transpose_input1 ? true : false;
     data->tB = params.transpose_input2 ? true : false;
     data->tI = params.transpose_input3 ? true : false;
-    data->k = input1_dims[params.transpose_input1 ? 1 : 0];
-    data->m = input1_dims[params.transpose_input1 ? 0 : 1];
-    data->n = input2_dims[params.transpose_input2 ? 1 : 0];
+    if (input1_dims[2] == 1 && input1_dims[3] == 1) {
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp]: change to if (input1_dims[2] & input1_dims[3]) {

File: amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp [hunk 4]
```diff
@@ -161,9 +177,20 @@ static vx_status VX_CALLBACK initialize(vx_node node, const vx_reference *parame
...
+        data->k = input1_dims[params.transpose_input1 ? 1 : 0];
+        data->m = input1_dims[params.transpose_input1 ? 0 : 1];
+    }
+    else if(input1_dims[0] == 1 && input1_dims[1] == 1) {
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp]: same as before

File: amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp [hunk 5]
```diff
@@ -161,9 +177,20 @@ static vx_status VX_CALLBACK initialize(vx_node node, const vx_reference *parame
...
+        data->k = input1_dims[params.transpose_input1 ? 3 : 2];
+        data->m = input1_dims[params.transpose_input1 ? 2 : 3];
+    }
+    if (input2_dims[2] == 1 && input2_dims[3] == 1) {
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp]: same as previous change

File: amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp [hunk 6]
```diff
@@ -161,9 +177,20 @@ static vx_status VX_CALLBACK initialize(vx_node node, const vx_reference *parame
...
+        data->n = input2_dims[params.transpose_input2 ? 1 : 0];
+    }
+    else if (input2_dims[0] == 1 && input2_dims[1] == 1) {
```
Comment (rrawther) [inline amd_openvx_extensions/amd_nn/src/tensor_matrix_multiply.cpp]: same as previous comment

File: model_compiler/python/nnir_to_openvx.py [hunk 7]
```diff
@@ -1727,6 +1754,16 @@ def generateTestCPP(graph,argmaxOutput,fileName,virtual_tensor_flag):
                             *dstG++ = src[1];
                             *dstB++ = src[0];
                         }
+                    } else if(data_type == VX_TYPE_INT64)
+                    {
+                        long int* dstR = (long int*)ptr + ((n * stride[3] + y * stride[1]) >> 3);
+                        long int* dstG = dstR + (stride[2] >> 2);
```
Comment (rrawther) [inline model_compiler/python/nnir_to_openvx.py]: double check the (stride >> 2). I think it need to be (stride >> 3)

File: model_compiler/python/nnir_to_openvx.py [hunk 8]
```diff
@@ -1806,6 +1851,16 @@ def generateTestCPP(graph,argmaxOutput,fileName,virtual_tensor_flag):
                             }
                         }
                     }
+                    else if (data_type == VX_TYPE_INT64) {
+                        long int * pc = (long int *)((short *)ptr + n * CHW + y * W + x);
```
Comment (rrawther) [inline model_compiler/python/nnir_to_openvx.py]: why casting pointer to (short *)? Shouldn't it be (long int *)

File: model_compiler/python/nnir_to_openvx.py [hunk 9]
```diff
@@ -2036,6 +2093,16 @@ def generateTestCPP(graph,argmaxOutput,fileName,virtual_tensor_flag):
                     if(!(err < maxError)) maxError = err;
                 }
             }
+            else if(data_type == VX_TYPE_INT64)
+            {
+                for(size_t i = 0; i < count; i++) {
+                    float src = (float)((long int*)ptr)[i];
```
Comment (rrawther) [inline model_compiler/python/nnir_to_openvx.py]: double check this statement. long int to float can result into undesired results

**Comment by LakshmiKumar23** [comment]
@kiritigowda ready to be merged

**Pull request merged**